src/Features/LanguageServer/Protocol/Handler/MapCode/MapCodeHandler.cs

```diff
@@ -0,0 +1,137 @@
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.MapCode;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler;
+
+[ExportCSharpVisualBasicStatelessLspService(typeof(MapCodeHandler)), Shared]
+[Method(WorkspaceMapCodeName)]
+internal sealed partial class MapCodeHandler : ILspServiceRequestHandler<MapCodeParams, LSP.WorkspaceEdit?>
+{
+    public const string WorkspaceMapCodeName = "workspace/mapCode";
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public MapCodeHandler()
+    {
+    }
+
+    public bool MutatesSolutionState => false;
+    public bool RequiresLSPSolution => true;
+
+    public async Task<WorkspaceEdit?> HandleRequestAsync(MapCodeParams request, RequestContext context, CancellationToken cancellationToken)
+    {
+        //TODO: handle request.Updates if not empty
+        Contract.ThrowIfNull(context.Solution);
+
+        using var _ = PooledDictionary<Uri, TextEdit[]>.GetInstance(out var uriToEditsMap);
+        foreach (var codeMapping in request.Mappings)
+        {
+            var mappingResult = await MapCodeAsync(codeMapping).ConfigureAwait(false);
+
+            // Assume no two codeMappings target a common document.
+            if (mappingResult is (Uri uri, TextEdit[] textEdits))
+                uriToEditsMap[uri] = textEdits;
+        }
+
+        // return a combined WorkspaceEdit
+        if (context.GetRequiredClientCapabilities().Workspace?.WorkspaceEdit?.DocumentChanges is true)
+        {
+            return new WorkspaceEdit
+            {
+                DocumentChanges = uriToEditsMap.Select(kvp => new TextDocumentEdit
+                {
+                    TextDocument = new OptionalVersionedTextDocumentIdentifier { Uri = kvp.Key },
+                    Edits = kvp.Value,
+                }).ToArray()
+            };
+        }
+        else
+        {
+            return new WorkspaceEdit
+            {
+                Changes = uriToEditsMap.ToDictionary(kvp => ProtocolConversions.GetDocumentFilePathFromUri(kvp.Key), kvp => kvp.Value)
+            };
+        }
+
+        async Task<(Uri, TextEdit[])?> MapCodeAsync(LSP.MapCodeMapping codeMapping)
+        {
+            var textDocument = codeMapping.TextDocument;
+            if (textDocument == null)
+                return null;
+
+            if (context.Solution.GetDocument(textDocument) is not Document document)
+                return null;
```

Comment:
I guess this could happen if the document being requested is in misc.  Consider logging here though, as any other scenario is unexpected and might indicate a race